### ISSUE TITLE
[QUESTION] Why is it unhealthy?

### ISSUE THREAD

**Issue by erseldev** [coreruleset/modsecurity-crs-docker#278]
I'm struggling to understand the reason for unhealthy states. I checked all the following images:

> services:
>   nginx:
>     # image: owasp/modsecurity-crs:3.3.5-nginx-alpine-202310170110  # healthy
>     # image: owasp/modsecurity-crs:3.3.5-nginx-202402140602  # unhealthy
>     # image: owasp/modsecurity-crs:4-nginx-alpine-202404070904  # unhealthy
>     # image: owasp/modsecurity-crs:4.3.0-nginx-alpine-202406090906  # unhealthy
>     # image: owasp/modsecurity-crs:4.5.0-nginx-202407300107  # unhealthy
>     # image: owasp/modsecurity-crs:4.5-nginx-202407300107  # unhealthy
>     # image: owasp/modsecurity-crs:4.5.0-nginx-alpine-202407300107 # unhealthy

I already have been using the "3.3.5-nginx-alpine-202310170110" without having any unhealthy status. Just wanted to upgrade the image and made some tests before doing so. Above I'm sharing the results for me.

What really has changed after "3.3.5-nginx-alpine-202310170110"?
What may be the reason?

BTW, the test site is loading without any issue with all the images above. Should I ignore the unhealthy status? I doubt it but just wanted to ask it anyway

Thanks ...

**Comment by theseion** [comment]
How are you checking the health status?

**Comment by erseldev** [comment]
I don't have any special directives. I'm using the default ones coming with the images.

**Comment by fzipi** [comment]
E.g. using the base docker compose file, and changing it for one of the versions mentioned here:
```
❯ docker ps
CONTAINER ID   IMAGE                          COMMAND                  CREATED         STATUS                             PORTS                                                 NAMES
680181ccf3bb   owasp/modsecurity-crs:apache   "/docker-entrypoint.…"   3 minutes ago   Up 29 seconds (health: starting)   80/tcp, 0.0.0.0:80->8080/tcp, 0.0.0.0:443->8443/tcp   modsecurity-crs-docker-crs-apache-1
❯ docker ps
CONTAINER ID   IMAGE                          COMMAND                  CREATED         STATUS                    PORTS                                                 NAMES
680181ccf3bb   owasp/modsecurity-crs:apache   "/docker-entrypoint.…"   3 minutes ago   Up 30 seconds (healthy)   80/tcp, 0.0.0.0:80->8080/tcp, 0.0.0.0:443->8443/tcp   modsecurity-crs-docker-crs-apache-1
```

**Comment by erseldev** [comment]
Thank you fzipi. I tried for longer times (more than 3 mins) with the same result.

**Comment by fzipi** [comment]
Weird. The [health check](https://github.com/coreruleset/modsecurity-crs-docker/blob/main/src/bin/healthcheck) is simple. Maybe it is the docker version? Don't know what to say :/

**Comment by erseldev** [comment]
With openresty I can reach to healthy status. However, now I have the following error with it:

> "ssl_stapling" ignored, issuer certificate not found for certificate "/usr/local/openresty/nginx/conf/server.crt"

I noticed I don't have the below entries in the openresty logs, while I'm having them in nginx logs:

nginx | /docker-entrypoint.sh: Launching /docker-entrypoint.d/10-generate-certificate.sh
nginx | /usr/local/bin/generate-certificate: generating new certificate

I'm going to stick to "3.3.5-nginx-alpine-202310170110" for some time. I'm happy with my old buddy... :)

**Comment by theseion** [comment]
That's just a warning you can ignore. We generate self-signed certificates so HTTPS works.

Can you maybe share a how you're running the image? Are you using Docker? `docker compose`? Which version of Docker? If we had more information, we might be able to figure this out.

**Comment by erseldev** [comment]
I'm using rootless docker on the following config:

> Client:
>  Version:    24.0.7
>  Context:    default
>  Debug Mode: false
>  Plugins:
>   buildx: Docker Buildx (Docker Inc.)
>     Version:  v0.16.1
>     Path:     /usr/libexec/docker/cli-plugins/docker-buildx
>   compose: Docker Compose (Docker Inc.)
>     Version:  2.24.6+ds1-0ubuntu1~22.04.1
>     Path:     /usr/libexec/docker/cli-plugins/docker-compose
> 
> Kernel Version: 5.15.0-118-generic
>  Operating System: Ubuntu 22.04.4 LTS
>  OSType: linux
>  Architecture: aarch64
>  CPUs: 8
>  Total Memory: 15.22GiB

**Comment by theseion** [comment]
Do you start the image with `docker run ...` or with `docker compose ...`? In either case, what does your command / compose config look like?

**Comment by erseldev** [comment]
I'm using  `docker compose up`  and `docker compose up --build`.

```

services:
  nginx:
    image: owasp/modsecurity-crs:3.3.5-nginx-alpine-202310170110  # healthy
    container_name: nginx
    volumes:
      - /home/ted/docker/proxy/conf.d/default.conf:/etc/nginx/templates/conf.d/default.conf.template
      - /home/ted/docker/sites/conf.d:/etc/nginx/conf.d
      - /home/ted/docker/sites/dhparam:/etc/nginx/dhparam:ro
      - /home/ted/docker/sites/letsencrypt/live:/etc/letsencrypt/live:ro
      - /home/ted/docker/sites/letsencrypt/archive:/etc/letsencrypt/archive:ro
      # temp volume for testing with simple index.html:
      - /home/ted/docker/sites/gpcom:/etc/nginx/sites/gpcom
      #todo django staticfiles, media volumes for each dj container:
      - gpcom_staticfiles:/etc/nginx/gpcom_staticfiles
      - gpcom_mediafiles:/etc/nginx/gpcom_mediafiles

    environment:
      # Required for OCSP Stapling - Use a different DNS server if desired (e.g. 8.8.8.8)
      - DNS_SERVER=127.0.0.11 1.1.1.1 1.0.0.1 [2606:4700:4700::1111] [2606:4700:4700::1001]
      - MODSEC_AUDIT_LOG_TYPE=Concurrent
      - MODSEC_AUDIT_LOG_PARTS=ABIJDEFHZ
    ports:
      - "80:80"
      - "443:443"
    user: "0:0"
    restart: always
    networks:
      - proxy
      - crowdsec
    logging:
      options:
        max-size: "50m"
        max-file: "3"

# creating site specific volumes:
volumes:
  gpcom_staticfiles:
    name: gpcom_staticfiles
  gpcom_mediafiles:
    name: gpcom_mediafiles

networks: # these are not used by this compose. they are created to make these networks ready upfront for the other services.
  proxy:
    driver: bridge
    name: proxy
  crowdsec:
    driver: bridge
    name: crowdsec
  socket-proxy:
    driver: bridge
    name: socket-proxy

```

This is a preliminary and simplified version and the one I'm playing around with. In the end, I will launch a django site and also going to hook crowdsec container to this django site.

Since django is going to bring a lot of complexities, I'm testing the waters with a simple index.html page to setup everything  properly on the server level.

There is a reference to socket-proxy. Socket-proxy is off as well. I'm playing around with only nginx/openresty as of now (not crowdsec either).

I setup Cloudflare side properly and all the other stuff like HSTS, OCSP, etc.. I'm receiving A+ on ssllabs.com tests.

**Comment by erseldev** [comment]
I forgot to say...
Letsencrypt Certs are managed by dnsrobocert which is also running in its own container....

**Comment by fzipi** [comment]
I see you have your own default template. Are you including the `nginx/templates/includes/location_common.conf.template` file there?

**Comment by erseldev** [comment]
No, I don't have it, but just made a search and figured that it is used to include healtz endpoint.

I've created the file as below and included as your suggestion:
```
location /healthz {
    access_log off;
    add_header Content-Type text/plain;
    return 200 "OK";
}
```
Tried the docker compose with image "owasp/modsecurity-crs:4.5.0-nginx-alpine-202407300107". Again unhealthy.

**Comment by fzipi** [comment]
And you confirmed the file is being included, right?

**Comment by erseldev** [comment]
I added it to the compose file like below:
      `- /home/ted/docker/proxy/conf.d/location_common.conf:/etc/nginx/templates/includes/location_common.conf.template`

After getting unhealthy status, removed this line from the compose file to bring it back to the original version as shared above. Now I'm getting the following error:

```
nginx  | 2024/08/15 21:10:54 [emerg] 1#1: "location" directive is not allowed here in /etc/nginx/conf.d/location_common.conf:1
nginx  | nginx: [emerg] "location" directive is not allowed here in /etc/nginx/conf.d/location_common.conf:1
```

I never had such an error before.

**Comment by fzipi** [comment]
One thing that you need to keep is the filesystem structure. Meaning that the location_common.conf file is included [inside the server configuration](https://github.com/coreruleset/modsecurity-crs-docker/blob/ab3f1e4eac5b1f75f13957bfb771d56c06882e43/nginx/templates/conf.d/default.conf.template#L32). How did you create your default.conf file? Based on [the one we provide](nginx/templates/conf.d/default.conf.template)?

**Comment by erseldev** [comment]
Hello again...
I made quite a bit of modifications on my side and placed the right structure as you suggested. Added some custom environment variables to achieve the maximum flexibility to support multiple sites on the same host.

After some hard work, everything is running great. Now I'm running the latest image: "4.5.0-nginx-alpine-202407300107".

To be frank, I was kind of reluctant to go through every detail again like I did some time ago. 
Thank you for pushing me for this direction.
